doc/src/guides/contributing/deprecations.md

```diff
@@ -0,0 +1,549 @@
+  users can get the basic idea of why the deprecation without having to read
+  through pages of old discussions. You may also give longer examples here
+  that would not fit in the {func}`~.sympy_deprecation_warning()` message or
+  `.. deprecated::` text.
+
+  Every deprecation should have a cross-reference target (using
+  `(target-name)=` above the section header) so that the `.. deprecated::`
+  note in the relevant docstring can refer to it. This target should also be
+  passed to the `active_deprecations_target` option of
+  {func}`~.sympy_deprecation_warning` or {func}`@deprecated
+  <sympy.utilities.decorator.deprecated>`. This will automatically put a link
+  to the page in the documentation in the warning message. The target name
+  should include the word "deprecation" or "deprecated" (target names are
+  global in Sphinx, so the target name needs to be unique across the entire
+  documentation).
+
+  The section header name should be the thing that is deprecated, and should
+  be a level 3 header under the corresponding version (typically it should be
+  added to the top of the file).
+
+  If multiple deprecations are related to one another, they can all share a
+  single section on this page.
+
+  If the deprecated function is not included in the top-level
+  `sympy/__init__.py` be sure to clearly indicate which submodule the object
+  is referring to. If you refer to anything that is documented in the Sphinx
+  module reference, cross-reference it, like ``` {func}`~.func_name` ```.
+
+  Note that examples here are helpful, but you generally should not use
+  doctests to show the deprecated functionality, as this will itself raise the
+  deprecation warning and fail the doctest. Instead you may use `# doctest:
+  +SKIP`, or just show the example as a code block instead of a doctest.
+
+  Here are examples corresponding to the (imaginary) examples above:
+
+  ```md
+  (simplify-this-deprecation)=
+  ### `simplify_this()`
+
+  The `sympy.simplify.simplify_this()` function is deprecated. It has been
+  replaced with the {func}`~.simplify` function. Code using `simplify_this()`
+  can be fixed by replacing `simplfiy_this(expr)` with `simplify(expr)`. The
+  behavior of the two functions is otherwise identical.
+
+  This change was made because `simplify` is a much more Pythonic name than
+  `simplify_this`.
+  ```
+
+  `````md
+  (is-this-zero-y-deprecation)=
+  ### `is_this_zero()` second argument
+  The second argument to {func}`~.is_this_zero()` is deprecated. Previously
+  `is_this_zero(x, y)` would check if x = y. However, this was removed because
+  it is trivially equivalent to `is_this_zero(x - y)`. Furthermore, allowing
+  to check $x=y$ in addition to just $x=0$ is is confusing given the function
+  is named "is this zero".
+
+  In particular, replace
+
+  ```py
+  is_this_zero(expr1, expr2)
+  ```
+
+  with
+
+  ```py
+  is_this_zero(expr1 - expr2)
+  ```
+
+  `````
+
```

Comment:
This page has a lot of discussion about when something should be deprecated. I think that it would be good if the examples used actually match that rationale i.e. neither of these is a good example. In the case of `simplify_this` there is no reason why the function couldn't just be left as a stub that calls `simplify` with a docstring saying that it's recommended to use `simplify` instead. Also in the case of `is_this_zero` there isn't any clear harm in leaving the two argument form there and just advising to use something else in the docs.

Reply:
That's fair, although both of these are very similar to real warnings that already exist. There are several of examples of existing deprecations that are just renames for clarity (e.g., `tensorhead` -> `tensor_heads`). Maybe we could say these weren't really worth it to begin with, though. 

The `is_this_zero` example is basically an inverted version of a current deprecation of `Eq` (`Eq(x)` with a single argument is deprecated). I honestly think this one is fine. It's a good example of something that is both confusing and completely unnecessary IMO.

I can definitely try to come up with better examples (suggestions welcome). Or reuse an existing real deprecation as an example, if it is simple enough (although I rather like how these are completely self-contained).

Reply:
I've lost count of the number of times I've copied snippets of code from somewhere and seen the deprecation warning about single argument Eq so it was definitely widely used. Having read the rationale for when to deprecate here (which I agree with) I don't think it was a reasonable thing to deprecate.

There's nothing inherently wrong with `Eq(x)` as a shorthand for `Eq(x, 0)` especially since in many contexts where an equation should logically be used sympy allows an expression with the implicit meaning that the expression is equal to zero. In that sense you can view `Eq(expr)` as simply a conversion between two different ways that sympy represents equations.

More importantly though nothing was gained by deprecating the single argument Eq. It didn't make it possible to make other useful changes or prevent the creation of nonsensical objects or anything else. It was just a deprecation for the sake of it. I agree that it would have been a cleaner API design not to allow single argument Eq in the first place but there was no real benefit from trying to change that retrospectively.

As for a better example I would go with something like `integrate(Eq(x, y), t)`. There's a clear rationale there for why it isn't possible to make that work sensibly. There is clear benefit in showing a message to users explaining why that code is probably incorrect. It isn't possible to just "fix the bug" because that requires introducing an integration constant and anyone currently depending on it won't expect that. If we add the symbol then that's incompatible with the rest of what integrate does and also we have to choose what symbol to use which should really be handled by the user. The code is broken and gives mathematically incorrect results in a way that cannot be fixed in general on SymPy's side without the user updating their code in some way. Ideally this should raise an error but since some code might depend on this we start with a deprecation warning.

Reply:
I think the boat's already sailed on single-argument Eq. Personally I think it was fine because Eq(x) is just as likely to be a typo, and explicitly writing Eq(x, 0) isn't hard. It's not unlike how a long time ago we used to make `__call__` do substitution but we removed it because people were more likely to mistype multiplication. 

I'll think if there are better examples that can be used here. Like I said I do rather like that these examples are small and self-contained, even if they are imaginary (people can look at the real deprecations if they want a more concrete example).